python/protobuf.c

```diff
@@ -0,0 +1,128 @@
+ *     * Neither the name of Google LLC nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL Google LLC BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "protobuf.h"
+
+#include "descriptor.h"
+#include "descriptor_pool.h"
+
+static void PyUpb_ModuleDealloc(void *module) {
+  PyUpb_ModuleState *s = PyModule_GetState(module);
+  upb_arena_free(s->obj_cache_arena);
+}
+
+static struct PyModuleDef module_def = {PyModuleDef_HEAD_INIT,
+                                        PYUPB_MODULE_NAME,
+                                        "Protobuf Module",
+                                        sizeof(PyUpb_ModuleState),
+                                        NULL,  // m_methods
+                                        NULL,  // m_slots
+                                        NULL,  // m_traverse
+                                        NULL,  // m_clear
+                                        PyUpb_ModuleDealloc};
+
+// -----------------------------------------------------------------------------
+// ModuleState
+// -----------------------------------------------------------------------------
+
+PyUpb_ModuleState *PyUpb_ModuleState_Get() {
+  PyObject *module = PyState_FindModule(&module_def);
+  return PyModule_GetState(module);
+}
+
+// -----------------------------------------------------------------------------
+// ObjectCache
+// -----------------------------------------------------------------------------
+
+void PyUpb_ObjCache_Add(const void *key, PyObject *py_obj) {
+  PyUpb_ModuleState *s = PyUpb_ModuleState_Get();
+  upb_inttable_insert(&s->obj_cache, (uintptr_t)key, upb_value_ptr(py_obj),
+                      s->obj_cache_arena);
+}
+
+void PyUpb_ObjCache_Delete(const void *key) {
+  PyUpb_ModuleState *s = PyUpb_ModuleState_Get();
+  upb_value val;
+  upb_inttable_remove(&s->obj_cache, (uintptr_t)key, &val);
+  assert(upb_value_getptr(val));
+}
+
+PyObject *PyUpb_ObjCache_Get(const void *key) {
+  PyUpb_ModuleState *s = PyUpb_ModuleState_Get();
+  upb_value val;
+  if (upb_inttable_lookup(&s->obj_cache, (uintptr_t)key, &val)) {
+    PyObject *ret = upb_value_getptr(val);
+    Py_INCREF(ret);
```

Comment:
where is the corresponding DECREF?

Reply:
This is a function that returns a reference to the caller when it succeeds. I added a comment describing this to the header file.